exporter/containerimage: default to oci-mediatypes=true

neersighted opened the pull request:
The change to provenance-by-default implicitly changed the default mediaTypes to OCI, but an explicit `--provenance=false` negates this and causes surprising behavior.

To fix this, we make the explicit default `oci-mediatypes=true` and make `oci-mediatypes=false` fail for incompatible cases.

Addresses https://github.com/moby/buildkit/issues/6094.

Tests are needed/missing; I've had some issues getting the tests working locally and skipped them for now.

neersighted committed:
exporter/containerimage: remove unused EnableForceCompression

This helper has not been used for some time.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
exporter/containerimage: default to oci-mediatype=true

To facilitate this change, we now fail when there is an exporter option
conflict instead of implicitly setting oci-mediatypes=true.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
exporter/containerimage: remove unnecessary errors.Errorf

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
exporter/containerimage: parse oci-mediatypes=bool strictly

This matches the documentation, which never specified behavior for
empty/unknown (non-boolean) strings.

We can also remove the parseBoolWithDefault helper, as this was the last
consumer.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted committed:
exporter/containerimage: document "oci-artifact=" values

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

tonistiigi commented on exporter/containerimage/opts.go:
```diff
@@ -49,7 +48,7 @@ func (c *ImageCommitOpts) Load(ctx context.Context, opt map[string]string) (map[
 		case exptypes.OptKeyName:
 			c.ImageName = v
 		case exptypes.OptKeyOCITypes:
-			err = parseBoolWithDefault(&c.OCITypes, k, v, true)
```
Not handling the empty value in here seems like possibly breaking change. Don't remember the history behind it but maybe don't break it as it seems harmeless.

neersighted replied:
That change was made for the other types that originally used this parsing function, so based on that I'm relatively confident in/in favor of this change.

tonistiigi reviewed (commented):
@neersighted Check the CI errors. Some tests expecting different mediatype and linter.

neersighted committed:
all: use well-known zstd mediaType constants

When support for zstd was introduced, these mediaTypes were not yet
available in released versions of their respective Go packages. That is
no longer the case.

Signed-off-by: Bjorn Neergaard <bjorn.neergaard@docker.com>

neersighted commented:
> Check the CI errors. Some tests expecting different mediatype and linter.
Yeah, I'm having issues getting the tests to run locally on arm64 due to platform mismatches in the Dockerfile linter tests throwing up a LOT of noise.

A lot of tests have implicit assumptions about this behavior, and I will work through them and make them pass (and add a regression test to ensure this behavior is as intended in the future) once I have a little more time to review the CI runs.

Unless you have some tips on how to get the tests suite to pass (on master) locally on an arm64 mac?

neersighted commented:
Just leaving a note that I still mean to get back to this and get the tests added/passing; but I've been having to deal with other concerns at work, and I'll be off for a few days. I'm going to block out some time to get this over the finish line when I get back.

tonistiigi commented:
@neersighted I'll carry this
